File: Problems/python/boj_연산자끼워넣기.py
#https://www.acmicpc.net/problem/14888
#DFS

n = int(input())
nums = list(map(int, input().split()))
op = list(map(int, input().split()))

minv = int(1e9) 
maxv = -int(1e9) 

def dfs(i, now, add, sub, mul, div):
    global minv, maxv
    if i==n:
        minv = min(now, minv)
        maxv = max(now, maxv)
        return
        
    if add:
        dfs(i+1, now+nums[i], add-1, sub, mul, div)
    if sub:
        dfs(i+1, now-nums[i], add, sub-1, mul, div)
    if mul:
        dfs(i+1, now*nums[i], add, sub, mul-1, div)
    if div:
        dfs(i+1, int(now/nums[i]), add, sub, mul, div-1)
        #음수를 나눌때는 //, %의 결과가 다르다

dfs(1, nums[0], op[0], op[1], op[2], op[3])

print(maxv)
print(minv)

File: Problems/cpp/boj_연산자끼워넣기.cpp
//https://www.acmicpc.net/problem/14888
//Brute-force, DFS

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

vector<int> op(4);
vector<int> vec;
int n, minv=1e9, maxv=-1e9;

void dfs(int depth, int num, int a, int s, int m, int d){
    if (depth==n){
        minv = min(minv, num);
        maxv = max(maxv, num);
        return;
    }

    if (a)
        dfs(depth+1, num+vec[depth], a-1, s, m, d);
    
    if (s)
        dfs(depth+1, num-vec[depth], a, s-1, m, d);
    
    if (m)
        dfs(depth+1, num*vec[depth], a, s, m-1, d);
    
    if (d)
        dfs(depth+1, num/vec[depth], a, s, m, d-1);
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);

    cin >> n;
    vec = vector<int>(n);
    for (int i=0; i<n; i++)
        cin >> vec[i];

    for (int i=0; i<4; i++)
        cin >> op[i];

    dfs(1, vec[0], op[0], op[1], op[2], op[3]);
    
    cout << maxv << '\n';
    cout << minv << '\n';
}